aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp

```diff
@@ -0,0 +1,115 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cpu/Loops.h>
+#include <ATen/native/quantized/cpu/quantized_ops.h>
+#include <ATen/quantized/Quantizer.h>
+#include <ATen/native/quantized/cpu/init_qnnpack.h>
+#include <ATen/native/quantized/cpu/qnnpack_utils.h>
+#include <c10/core/TensorOptions.h>
+#include <caffe2/utils/threadpool/ThreadPoolMobile.h>
+
+#include <algorithm>
+
+namespace at {
+namespace native {
+
+namespace {
+Tensor quantized_channel_shuffle_impl(
+    const Tensor& self,
+    int64_t groups) {
+
+  TORCH_CHECK(
+      groups > 0,
+      "Number of groups to divide channels in must be positive.",
+      " Value of groups:", groups);
+  TORCH_CHECK(
+      self.dim() == 4,
+      "channel_shuffle expects 4D input, but got input with sizes ",
+      self.sizes());
+  TORCH_CHECK(
+      self.scalar_type() == kQUInt8,
+      "Quantized channel shuffle works only on uint8_t.",
+      "But got:", self.scalar_type());
+  const Tensor self_nhwc = self.contiguous(MemoryFormat::ChannelsLast);
+  Tensor qy = at::new_qtensor_cpu(
+      self_nhwc.sizes(),
+      TensorOptions(kQUInt8)
+          .device(kCPU)
+          .memory_format(MemoryFormat::ChannelsLast),
+      make_per_tensor_affine_quantizer(
+          self_nhwc.q_scale(), self_nhwc.q_zero_point(), kQUInt8)
+      );
+
+  // Degenerate case of just copying.
+  if (groups == 1) {
+    qy.copy_(self_nhwc);
+    return qy.contiguous(self.suggest_memory_format());
+  }
+
+  int64_t channels = self.size(1);
+  TORCH_CHECK(channels > 0,
+             "Number of channels must be positive, got:", channels);
+  TORCH_CHECK((channels % groups) == 0,
+             "Number of channels must be divisible gy groups. Got ",
+             channels, " channels and ", groups, " groups.");
+
+  initQNNPACK();
+
+  pytorch_qnnp_operator_t qnnpack_operator{nullptr};
+
+  const pytorch_qnnp_status createStatus = pytorch_qnnp_create_channel_shuffle_nc_x8(
+      groups /* groups */,
+      channels / groups /* group channels */,
+      0 /* flags */,
+      &qnnpack_operator);
+  TORCH_INTERNAL_ASSERT(
+      createStatus == pytorch_qnnp_status_success,
+      "failed to create QNNPACK ChannelShuffle operator");
+
+  std::unique_ptr<pytorch_qnnp_operator, QnnpackOperatorDeleter>
+      qnnpack_uniq_ptr(qnnpack_operator);
```

Comment:
I commented on the old diff as well, but this style ensure that the operator is freed even if we throw an exception.